### DISCUSSION TITLE
How to build wasm on Termux?

### DISCUSSION THREAD

**Discussion by ghost** [framesurge/perseus#206] [General]
Termux is a terminal emulator for Android, my CPU architecture is ARM64. Both "perseus build" and "cargo build" failed, and have the same error:
error: expected identifier or integer
 --> /home/alarm/hello/src/lib.rs:9:21
  |
9 |                 p { "Hello World!" }
While using "wasm-pack build", it got many errors, for example:
error[E0514]: found crate `core` compiled by an incompatible version of rustc

**Comment by arctic-hen7** [comment]
Ah! I actually develop Perseus on my tablet occasionally and I have encountered so many bugs with it! Make sure you've got the wasm Rust package (I think it's `rust-std-wasm32`, but a quick `apt` search should find it), and then everything should work.

If you've already got that package, then maybe try v0.4.x (wasm-pack sounds like v0.3.x)? Also make sure you have the latest Rust package on your device.

**Reply by ghost** [reply]
> If you've already got that package, then maybe try v0.4.x (wasm-pack sounds like v0.3.x)? Also make sure you have the latest Rust package on your device.

Couldn't find wasm-pack. The "apt search wasm" command only found these:
rust-nightly-wasm32-unknown-unknown/stable,stable 1.61.0-1 all
  rust std for wasm32-unknown-unknown target

rust-std-wasm32/stable,stable 1.64.0 all
  rust std for wasm32-unknown-unknown target

wasmedge/stable 0.11.1 aarch64
  A lightweight, high-performance, and extensible WebAssembly runtime

wasmer/stable 2.3.0 aarch64
  A fast and secure WebAssembly runtime

I used Arch Linux ARM in Termux to install wasm-pack, Termux can install some distributions, but Termux itself doesn't provide wasm-pack.

**Reply by arctic-hen7** [reply]
Install `rust-std-wasm32`, that will give you the ability to compile for Wasm. Then `cargo install wasm-pack` (only for v0.3.x).

**Reply by ghost** [reply]
> Then `cargo install wasm-pack` (only for v0.3.x).

Why not develop perseus to use the latest wasm-pack? Wasm-pack now has version 0.10.3

**Reply by ghost** [reply]
The perseus version is 0.3.6, the content in lib.rs is:
use perseus::{Html, PerseusApp, Template};
use sycamore::view;

#[perseus::main]
pub fn main<G: Html>() -> PerseusApp<G> {
    PerseusApp::new().template(|| {
        Template::new("index").template(|_| {
            view! {
                p { "Hello World!" }
            }
        })
    })
}

The wasm-pack version is 0.3.1, but after running "perseus build":
perseus build
⠄ [1/2] 🔨 Generating your app...                         [2/2] 🏗️  Building your app to Wasm...❌
error: Found argument 'build' which wasn't expected, or isn't valid in this context                             
USAGE:                                                      wasm-pack [FLAGS] <SUBCOMMAND>
  [1/2] 🔨 Generating your app...❌                       [2/2] 🏗️  Building your app to Wasm...❌
   Compiling hello v0.1.0 (/home/alarm/hello)           error: expected identifier or integer
 --> /home/alarm/hello/src/lib.rs:9:21
  |                                                     9 |                 p { "Hello World!" }
  |                     ^^^^^^^^^^^^^^

**Reply by arctic-hen7** [reply]
No no, I meant _Perseus_ v0.3.x. In Perseus v0.4.x, wasm-pack is no longer needed at all. Install the latest version of wasm-pack and it should work.

**Comment by ghost** [comment]
Can I use QEMU to emulate x86 to compile Rust to wasm? The mobile phone only ships with ARM CPU, will QEMU help?

**Reply by arctic-hen7** [reply]
This shouldn't be necessary. Wasm-pack is a Cargo package, you don't need to install it through Termux's package management.
